### PR TITLE
[WIP]EOS-19081 : Implementation of insert, delete, lookup operation for Btree

### PR DESCRIPTION
**Pull request by RadhaGulhane13**
Signed-off-by: RadhaGulhane13 <radha.gulhane@seagate.com>

### CODE REVIEW

**Comment by RadhaGulhane13** [comment]
Things which are not yet implemented or needs to be discussed :
- Condition for overflow, underflow
- size of node while allocating new node if overflow : ref. line -> 717
- assign of new key value to bop->bo_rec  : ref. line -> 747-748

File: btree/btree.c [hunk 1]
```diff
@@ -651,6 +651,174 @@ static int get_tick(struct m0_btree_op *bop)
...
+				if(bop->bo_rec.key < slot_.s_rec.key) {
+					node_make (&slot, bop->bo_tx);
+					node_set(&slot, bop->bo_tx);
+				} else {
+					node_find(&slot_,bop->bo_rec.key);
+					slot_.s_rec = bop->bo_rec;
+					node_make (&slot_, bop->bo_tx);
+					node_set(&slot_, bop->bo_tx);
+				}
+				struct slot slot_ = {
```
Comment (shashank-parulekar) [inline btree/btree.c:777]: You cannot have the same named variable (slot_) in the same scope.

File: btree/btree.c [hunk 2]
```diff
@@ -651,6 +651,174 @@ static int get_tick(struct m0_btree_op *bop)
...
+			};
+			node_find(&slot,bop->bo_rec.r_key);
+			level->l_idx = slot.idx;
+
+			if (node_isfit(&slot)) {
+				node_make (&slot, bop->bo_tx);
+				return ACT;
+			} else {
+				node_move(level->l_node, level->l_alloc, D_RIGHT, NR_EVEN, bop->bo_tx);
+				struct slot slot_ = {
```
Comment (shashank-parulekar) [inline btree/btree.c:763]: If possible use some better name maybe 'slot_for_right_node' so that code reading becomes easier.

File: btree/btree.c [hunk 3]
```diff
@@ -651,6 +651,174 @@ static int get_tick(struct m0_btree_op *bop)
...
+					slot_.s_rec = bop->bo_rec;
+					node_make (&slot_, bop->bo_tx);
+					node_set(&slot_, bop->bo_tx);
+				}
+				struct slot slot_ = {
+				.s_node = level->l_alloc;
+				.s_idx  = 0;
+				};
+				bop->bo_rec.r_key = node_key(&slot_);
+				bop->bo_rec.r_value = l_alloc; //update bo_rec.r_value
```
Comment (shashank-parulekar) [inline btree/btree.c:782]: This l_alloc needs to be another allocated node. Need to see if we saved a list of these allocated nodes at some other location.

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

File: btree/btree.c [hunk 4]
```diff
@@ -651,6 +651,174 @@ static int get_tick(struct m0_btree_op *bop)
...
+			return node_alloc(&node_op, tree, size,
+				oi->i_level[level].l_node->n_type, bop->bo_tx, P_ALLOC);
+		} else
+			return P_LOCK;
+	case P_LOCK:
+		if (!locked)
+			return lock_op_init(&bop->bo_op, &bop->bo_i->i_lop,
+					    P_CHECK);
+		else
+			return P_CHECK;
```
[the remark this answers is not in the document]
Comment (RadhaGulhane13) [reply btree/btree.c:726]: I think ,here , we are checking if there is already lock by the operation if not we are going for lock_op_init(&bop->bo_op, &bop->bo_i->i_lop, P_CHECK); else going for P_CHECK

and wait for lock release is done by lock_op_init function when tree is lock and we want lock for that particular operation

**Comment by RadhaGulhane13** [comment]
State diagram of insertion should get changed, as we are first reaching to leaf node and then performing allocation if needed and going one level up or else to P_LOCK,
the part of state diagram :
+------+
|           |
|           v
+-----NEXTDOWN--->LOCK
   ^       |
    |       v
   +-ALLOC

should change to :

+---+
|      v
+--NEXTDOWN--> ALLOC-->LOCK
             ^     |
             +---+

File: btree/btree.c [hunk 5]
```diff
@@ -651,6 +651,480 @@ static int get_tick(struct m0_btree_op *bop)
...
+	case P_LOCKALL:
+		if (bop->bo_flags & OF_LOCKALL)
+			return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN);
+	case P_DOWN:
+		oi->i_used = 0;
+		/* Load root node. */
+		return node_get(&oi->i_nop, tree, &tree->t_root, P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    slot = {};
```
Comment (ydb242) [inline btree/btree.c:1009]: It is bad practice to keep the struct name and its variable's name the same.
Please update it.

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

**Comment by huanghua78** [comment]
Hello @RadhaGulhane13 , is this PR still needed?
If yes, please address the review comments.
Otherwise, please close this PR.

**Comment by RadhaGulhane13** [comment]
> Hello @RadhaGulhane13 , is this PR still needed?
> If yes, please address the review comments.
> Otherwise, please close this PR.

Yes @huanghua78 , this PR is still in progress. I have working on comments posted 23 April, also doing some other changes , Once it is done will push the code.

File: btree/btree.c [hunk 6]
```diff
@@ -738,67 +753,105 @@ static int get_tick(struct m0_btree_op *bop)
...
-			struct slot slot = {
-			.s_node = level->l_node;
-			.s_rec = bop->bo_rec;
-			};
-			node_find(&slot,bop->bo_rec.r_key);
-			level->l_idx = slot.idx;
+	case P_MAKESPACE: {
+		struct slot slot_for_left_node = {
+		.s_node = level->l_node;
+		.s_rec = bop->bo_rec;
```
[the remark this answers is not in the document]
Comment (RadhaGulhane13) [reply btree/btree.c:759]: okay

File: btree/btree.c [hunk 7]
```diff
@@ -738,67 +753,105 @@ static int get_tick(struct m0_btree_op *bop)
...
+		if (node_isfit(&slot_for_left_node)) {
+			node_make (&slot_for_left_node, bop->bo_tx);
+			return ACT;
+		} else {
+			node_move(level->l_node, level->l_alloc, D_RIGHT, NR_EVEN, bop->bo_tx);
+			struct slot slot_for_right_node = {
+			.s_node = level->l_alloc;
+			.s_idx  = 0;
+			};
+			struct slot *tgt;
```
[the remark this answers is not in the document]
Comment (RadhaGulhane13) [reply btree/btree.c:772]: okay

File: btree/btree.c [hunk 8]
```diff
@@ -665,32 +672,34 @@ static int get_tick(struct m0_btree_op *bop)
 		else
 			return P_SETUP;
 	case P_COOKIE:
-		if (cookie_is_valid(tree, &bop->bo_key.k_cookie))
+		if (cookie_is_valid(tree, &bop->bo_key.k_cookie) &&
+		    node_space(k_cookie.node)> about_to_OVERFLOW))
```
[the remark this answers is not in the document]
Comment (RadhaGulhane13) [reply btree/btree.c:676]: will do the changes

**Pull request closed without merging**